src/cockpit/samples.py

```diff
@@ -1,65 +1,180 @@
-    'cpu.basic.iowait': ('millisec', 'counter', False, CPU_SAMPLER),
-
-    'cpu.core.nice': ('millisec', 'counter', True, CPU_SAMPLER),
-    'cpu.core.user': ('millisec', 'counter', True, CPU_SAMPLER),
-    'cpu.core.system': ('millisec', 'counter', True, CPU_SAMPLER),
-    'cpu.core.iowait': ('millisec', 'counter', True, CPU_SAMPLER),
+# This file is part of Cockpit.
+#
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-    'memory.free': ('bytes', 'instant', False, MEMORY_SAMPLER),
-    'memory.used': ('bytes', 'instant', False, MEMORY_SAMPLER),
-    'memory.cached': ('bytes', 'instant', False, MEMORY_SAMPLER),
-    'memory.swap-used': ('bytes', 'instant', False, MEMORY_SAMPLER),
-}
-
-
-# Friendly utility class to allow dynamically building the tree
-class Samples(dict):
-    def __getitem__(self, k):
-        if k not in self:
-            self[k] = Samples()
-        return self.get(k)
+import os
+import os.path
+from typing import Any, Dict, List, NamedTuple
 
 
 USER_HZ = os.sysconf(os.sysconf_names['SC_CLK_TCK'])
 MS_PER_JIFFY = 1000 / (USER_HZ if (USER_HZ > 0) else 100)
 
 
-def cockpit_memory_samples(samples):
-    with open('/proc/meminfo') as meminfo:
-        items = {k: int(v.strip(' kB\n')) for line in meminfo for k, v in [line.split(':', 1)]}
-
-    samples['memory.free'] = 1024 * items['MemFree']
-    samples['memory.used'] = 1024 * (items['MemTotal'] - items['MemAvailable'])
-    samples['memory.cached'] = 1024 * (items['Buffers'] + items['Cached'])
-    samples['memory.swap-used'] = 1024 * (items['SwapTotal'] - items['SwapFree'])
-
-
-def cockpit_cpu_samples(samples):
-    with open('/proc/stat') as stat:
-        for line in stat:
-            if not line.startswith('cpu'):
-                continue
-            cpu, user, nice, system, _idle, iowait = line.split()[:6]
-            core = cpu[3:] or None
-            prefix = 'cpu.core' if core else 'cpu.basic'
-
-            samples[f'{prefix}.nice'][core] = int(nice) * MS_PER_JIFFY
-            samples[f'{prefix}.user'][core] = int(user) * MS_PER_JIFFY
-            samples[f'{prefix}.system'][core] = int(system) * MS_PER_JIFFY
-            samples[f'{prefix}.iowait'][core] = int(iowait) * MS_PER_JIFFY
-
-
-s = Samples()
-cockpit_memory_samples(s)
-cockpit_cpu_samples(s)
-
-print(s)
+class SampleDescription(NamedTuple):
+    units: str
```

Comment:
Now that we're just doing a linear scan of all of the descriptions anyway, why not make the name part of the description, and turn the whole thing into a list?

ie, not:
```python
{
    'a.b.c': Description('ms', ...),
    'd.e.f': Description('ms', ...),
}
```

but

```python
[
    Description('a.b.c', 'ms', ...),
    Description('d.e.f', 'ms', ...),
]
```

This reads a bit nicer to me, and having the name built into the description will better match `MetricInfo` and help you construct those (see comment above).